app/routines/update_user_salesforce_info.rb

```diff
@@ -0,0 +1,148 @@
+class UpdateUserSalesforceInfo
+  USER_BLOCK_SIZE = 1000
+
+  def initialize
+    @errors = []
+  end
+
+  def self.call
+    new.call
+  end
+
+  def call
+    contacts_by_email = {}
+    contacts_by_id = {}
+
+    # Store a map of contacts by primary email
+
+    contacts.each do |contact|
+      next if contact.email.nil?
+      contacts_by_email[contact.email] = contact
+      contacts_by_id[contact.id] = contact
+    end
+
+    # Add contacts by alt email if they don't already exist in the map; error if they do.
+
+    contacts.each do |contact|
+      next if contact.email_alt.nil?
+
+      if contacts_by_email[contact.email_alt].present?
+        error!(message: "#{contact.email_alt} is an alt email on contact #{contact.id} but a " \
+                        "primary email on contact #{contacts_by_email[contact.email_alt].id}")
+      else
+        contacts_by_email[contact.email_alt] = contact
+        contacts_by_id[contact.id] = contact
+      end
+    end
+
+    # Go through all users that have already have a Salesforce ID and make sure
+    # their SF information is stil the same.
+
+    User.where("salesforce_contact_id IS NOT NULL")
+        .find_each do |user|
+
+      begin
+        contact = contacts_by_id[user.salesforce_contact_id]
+        cache_contact_data_in_user(contact, user)
+        user.save! if user.changed?
+      rescue StandardError => ee
+        error!(exception: ee, user: user)
+      end
+    end
+
+    # Go through all users that don't yet have a Salesforce ID and populate their
+    # salesforce info when they have verified emails that match SF data.
+
+    sf_emails = contacts_by_email.keys
+
+    User.eager_load(:contact_infos)
```

Comment:
I suggest also adding a `joins(:contact_infos)` to force an inner join. Might not actually matter.